src/lsm/set_associative_cache.zig

```diff
@@ -0,0 +1,423 @@
+        tags: []Tag,
+        values: []align(value_alignment) Value,
+        counts: []u64,
+        clocks: []u64,
+
+        pub fn init(allocator: mem.Allocator, value_count_max: u64) !Self {
+            assert(math.isPowerOfTwo(value_count_max));
+            assert(value_count_max > 0);
+            assert(value_count_max >= layout.ways);
+            assert(value_count_max % layout.ways == 0);
+
+            const sets = @divExact(value_count_max, layout.ways);
+            assert(math.isPowerOfTwo(sets));
+
+            const value_size_max = value_count_max * @sizeOf(Value);
+            assert(value_size_max >= layout.cache_line_size);
+            assert(value_size_max % layout.cache_line_size == 0);
+
+            const counts_size = @divExact(value_count_max * layout.clock_bits, 8);
+            assert(counts_size >= layout.cache_line_size);
+            assert(counts_size % layout.cache_line_size == 0);
+
+            const clocks_size = @divExact(sets * clock_hand_bits, 8);
+            assert(clocks_size >= layout.cache_line_size);
+            assert(clocks_size % layout.cache_line_size == 0);
+
+            const tags = try allocator.alloc(Tag, value_count_max);
+            errdefer allocator.free(tags);
+
+            const values = try allocator.allocAdvanced(
+                Value,
+                value_alignment,
+                value_count_max,
+                .exact,
+            );
+            errdefer allocator.free(values);
+
+            const counts = try allocator.alloc(u64, @divExact(counts_size, @sizeOf(u64)));
+            errdefer allocator.free(counts);
+
+            const clocks = try allocator.alloc(u64, @divExact(counts_size, @sizeOf(u64)));
+            errdefer allocator.free(clocks);
+
+            _ = Clocks;
+            _ = Counts;
+
+            var self = Self{
+                .sets = sets,
+                .tags = tags,
+                .values = values,
+                .counts = counts,
+                .clocks = clocks,
+            };
+
+            self.reset();
+
+            return self;
+        }
+
+        pub fn deinit(self: *Self, allocator: mem.Allocator) void {
+            assert(self.sets > 0);
+            self.sets = 0;
+
+            allocator.free(self.tags);
+            allocator.free(self.values);
+            allocator.free(self.counts);
+            allocator.free(self.clocks);
+        }
+
+        pub fn reset(self: *Self) void {
+            mem.set(Tag, self.tags, 0);
+            mem.set(u64, self.counts, 0);
+            mem.set(u64, self.clocks, 0);
+        }
+
+        pub fn get(self: *Self, key: Key) ?*align(value_alignment) Value {
+            const set = self.associate(key);
+            return self.search(set, key);
+        }
+
+        pub const GetOrPutResult = struct {
+            value_ptr: *align(value_alignment) Value,
+            found_existing: bool,
+        };
+
+        pub fn get_or_put(self: *Self, key: Key) GetOrPutResult {
```

Comment:
For the object cache, we also need an unconditional put method that doesn't carry the additional overhead of a search and that doesn't need to make any function calls or assertions related to locking.

That would be for cases where we just want to update the latest value.